### ISSUE TITLE
Suggestion: Group Methods or Classes

### ISSUE THREAD

**Issue by norswap** [Kotlin/dokka#84]
Would it be possible to create groups of methods and/or classes in a package, for the purpose of documentation?

Rationale: the documentation pages can get quite long, and there is often a logical grouping to be made. At the same time, one does not necessarily want to split the methods between packages and require extra imports.

Being able to require that top-level methods and classes be grouped per file for certain files (maybe specify this as a file-scope annotation?) would be a great start.

**Comment by yole** [comment]
If there is a logical grouping of entities in your code, why deprive those who work with your code from access to that grouping, and present it only in the documentation?

I don't see how, with IntelliJ's import management features, requiring extra imports is a problem at all. You don't need to write them, and you don't need to see them.

**Comment by norswap** [comment]
> If there is a logical grouping of entities in your code, why deprive those who work with your code from access to that grouping, and present it only in the documentation?

They wouldn't be deprived of it, if the methods are grouped inside the same file, or appear sequentially in the code separated by some form of heading comment (e.g., `// ----- My Section ------`).

> I don't see how, with IntelliJ's import management features, requiring extra imports is a problem at all. You don't need to write them, and you don't need to see them.

There is a cognitive dissonance here with the fact that everything in the standard library is imported by default (unlike in Java).

----

For an example of why this would be beneficial, consider https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-list/. I don't consider the laundry list of methods at the bottom to be especially user friendly. It would pay to split it a few sections: one for accessing elements of the list, one for filtering or taking subsets of the list, one for conversions, one for sorting, ...

**Comment by superbobry** [comment]
Here's a [good example](http://docs.scipy.org/doc/numpy/reference/routines.math.html) of how useful grouping can be for libraries with large number of functions. It isn't from the Java world, but it illustrates the point nonetheless.

**Comment by yole** [comment]
They wouldn't be deprived of it if they looked at the source code of your library. However, they wouldn't see any of this structure if they used regular IDE features such as code completion to explore its API.

The main reason why most of the standard library is imported by default (actually not everything - e.g. `kotlin.comparators` isn't imported by default) is that the split of the standard library into packages happened very late in the development of Kotlin 1.0, and we wanted to minimize the impact on existing applications.

I agree that the list of extension functions for `List` is kind of hard to read, but given that those extension functions are declared in many different places, a file-based grouping wouldn't help much. A better solution would be a tag such as `@group` that could be applied to the doc comments of different functions, regardless of where they are declared.

**Comment by aajtodd** [comment]
any status on this? Is it currently possible as a custom plugin?

**Comment by kamildoleglo** [comment]
Yes, certainly possible with a custom plugin, however no work has been done

**Comment by vincentsong** [comment]
Also looking for this feature

**Comment by aajtodd** [comment]
Is there an example or any direction on how we could do this with a custom plugin?

**Comment by IgnatBeresnev** [comment]
Hi @aajtodd 

Dokka has some documentation on plugin development that can be found here: https://kotlin.github.io/dokka/1.6.10/developer_guide/introduction/

As for example of how to group methods/classes together, I think it depends on what exactly you want to do, but you can have a look at [SameMethodNamePageMergerStrategy](https://github.com/Kotlin/dokka/blob/c44bf5487bd32f90a4576859548f1db0e9355a07/plugins/base/src/main/kotlin/transformers/pages/merger/SameMethodNamePageMergerStrategy.kt). As can be concluded from the name, it merges methods that have the same name and puts them on the same page :) An example of such page for kotlinx.serialization [here](http://dokka-snapshots.s3.eu-central-1.amazonaws.com/master/serialization/8537f0f/kotlinx-serialization-core/kotlinx.serialization/serializer-or-null.html)

**Comment by aaccioly** [comment]
Just dropping this here in case it generates some good will towards categories. @donraab has  written a great book about the subject (see: https://donraab.medium.com/book-eclipse-collections-categorically-37591b2135f5), he and the community also did a great job categorising parts of the Eclipse Collection API (see https://donraab.medium.com/categorizing-the-methods-of-set-types-in-eclipse-collections-d6d7d8dfe435). IMO, this is really useful for developers. Also, apparently some folks in the Groovy community are exploring tooling support for it as well: https://issues.apache.org/jira/browse/GROOVY-11482

I'll try to get myself familiarised with Dokka's plugin API. Hopefully there's an easy way to introduce, say a `Category` or `Group` annotation and respective transformers to group, link and diplay "categories" as Donald Raab is doing with IntelliJ's custom folding regions and custom html in Javadoc:

https://donraab.medium.com/grouping-java-methods-using-custom-code-folding-regions-with-intellij-bdff0b4fb2a3

**Comment by aaccioly** [comment]
Another possible workaround or idea. Javadoc has `@Documented`, Kotlin has `@MustBeDocumented`, so as a first effort, without changing the tooling, one can create annotations for each category group. The annotations will then show up in the docs next to each method, extension function, etc.

With the `-use` flag, Javadoc is also able to generate reverse lookup pages like this one: https://docs.oracle.com/en/java/javase/26/docs/api/java.base/java/lang/class-use/SafeVarargs.html. I'm not sure if Dokka itself or the dokka-javadoc plugin can generate equivalent "usage" pages (AFAIK there's no equivalent yet). With Javadoc, the combination of `@Documented` annotations plus usage pages can produce something close to Smalltalk-like method categories. Still not as visible, usable, or discoverable as what can be theoretically done with a plugin, but it is a possible start for an "right now" solution.
